File: built-in plugins dsp/mb-strings-tremolo.cpp
/**
 * MB Tremolo Strings
 * Category : instrument
 * Type     : strings
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Dramatic tremolo strings
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STRINGS_TREMOLO_H
#define MB_STRINGS_TREMOLO_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStringsTremolo : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-strings-tremolo";
    static constexpr const char* PLUGIN_NAME    = "MB Tremolo Strings";
    static constexpr const char* PLUGIN_TYPE    = "strings";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float speed = 8f;  // range [4, 16]
    float volume = 0.8f;  // range [0, 1]
    };

    MbStringsTremolo() = default;
    ~MbStringsTremolo() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.speed = std::clamp(params.speed, 4f, 16f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Tremolo Strings
        return input;
    }
};

#endif // MB_STRINGS_TREMOLO_H

File: built-in plugins dsp/mb-strings-tremolo.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbStringsTremoloPlugin: PluginDefinition = {
    id: 'mb-strings-tremolo', slug: 'mb-strings-tremolo', name: 'MB Tremolo Strings', category: 'instrument', type: 'strings', version: '1.0.0',
    description: 'Dramatic tremolo strings', author: 'Max Booster',
    oscillators: [{ type: 'sawtooth', detune: -3, gain: 0.35 }, { type: 'sawtooth', detune: 3, gain: 0.35 }, { type: 'triangle', detune: 0, gain: 0.3 }],
    envelope: { attack: 0.02, decay: 0.2, sustain: 0.9, release: 0.3 },
    parameters: [
      { id: 'speed', name: 'Speed', type: 'float', defaultValue: 8, minValue: 4, maxValue: 16, automatable: true },
      { id: 'volume', name: 'Volume', type: 'float', defaultValue: 0.8, minValue: 0, maxValue: 1, automatable: true },
    ],
    defaultPreset: { speed: 8, volume: 0.8 },
  };

export default MbStringsTremoloPlugin;
